bip-0322.mediawiki

```diff
@@ -13,119 +13,99 @@
+Additionally, the current message signing only proves that the message has been committed to by the recipient of a given invoice address.
+It does not prove anything about the invoice address itself, nor that the signer has access to the private keys used to implement this invoice.
+More importantly, it does not prove ownership nor access to any funds, even if the same private key would be a valid signer for spending them - and this is a commonly desired use case.
 
-=== SignatureProof container ===
+== Specification ==
 
-{|class="wikitable" style="text-align: center;"
-|-
-!Type
-!Length
-!Name
-!Comment
-|-
-|VarInt||1-8||scriptsiglen||Number of bytes in scriptSig data
-|-
-|Uint8*||[scriptsiglen]||scriptsig||ScriptSig data
-|-
-|VarInt||1-8||witlen||Number of entries in witness stack
-|-
-|Uint8[]*||[witlen]||wit||Witness stack, as [witlen] uint8* vectors, each one prepended with a varint of its size
-|}
+This BIP follows the specification of BIP-325 challenges and solutions (see Signet comparison below).
 
-In some cases, the scriptsig or wit may be empty. If both are empty, the proof is incomplete.
+Let there be two virtual transactions to_spend and to_sign.
 
-=== Result Codes ===
+The "to_spend" transaction is:
 
-A verification call will return a result code according to the table below.
+    nVersion = 0
+    nLockTime = 0
+    vin[0].prevout.hash = 0000...000
+    vin[0].prevout.n = 0xFFFFFFFF
+    vin[0].nSequence = 0
+    vin[0].scriptSig = OP_0 PUSH32[ message_hash ]
+    vin[0].scriptWitness = []
+    vout[0].nValue = 0
+    vout[0].scriptPubKey = message_challenge
 
-{|class="wikitable" style="text-align: center;"
-|-
-!Code
-!Description
-|-
-|INCOMPLETE||Empty proof.
-|-
-|INCONCLUSIVE||The given proof was consensus-valid but policy-invalid.
-|-
-|VALID||The proof was valid.
-|-
-|INVALID||The proof was invalid
-|-
-|ERROR||An error was encountered
-|}
+where message_hash is a BIP340-tagged hash of the message, i.e. sha256_tag(m), where tag = "BIP0322-signed-message", and message_challenge is the to be proven (public) key script.
+For proving funds, message_challenge shall be simply OP_TRUE.
 
-== Signing and Verifying ==
+The "to_sign" transaction is:
 
-If the challenge consists of an address is in the P2PKH (legacy) format, sign using the legacy format (further information below). Otherwise continue as stated below.
+    nVersion = 0 or as appropriate (e.g. 2, for time locks)
+    nLockTime = 0 or as appropriate (for time locks)
+    vin[0].prevout.hash = to_spend.txid
+    vin[0].prevout.n = 0
+    vin[0].nSequence = 0 or as appropriate (for time locks)
+    vin[0].scriptWitness = message_signature
+    vout[0].nValue = 0
+    vout[0].scriptPubKey = OP_RETURN
 
-For both cases, generate a sighash based on the given scriptPubKey and message as follows:
+When a proof of funds is being created, additional inputs should be included for virtually spending transaction outputs of desired value.
```

Comment:
MUST is correct, thanks.

Reply:
Well, to turn it around, if I give you a proof and it has inputs, it's a proof of funds for those inputs. If you ask me for a proof of funds for input A and I give you one for input B, it's the same as if I gave you a proof for bc1qA when you asked for bc1qB.

In that sense, I think "should" is correct, since it's up to the prover to put in what the verifier is asking for.

Reply:
Still a bit on the fence for this one. Obviously, the challenged does not have to fulfill a challenge, but responding with other information instead seems a bit pointless. :)